### DISCUSSION TITLE
At "Testing I" step and LS2 won't light

### DISCUSSION THREAD

**Discussion by bohnertchris** [thehighnibble/thehighnibble.github.io#120] [Cromemco Z-1 replica]
I have just finished step 26 in the assembly instructions (i.e. all ICs are in place). "Testing I" almost succeeds entirely. The ESP32-PICO-KIT starts properly as indicated by its terminal output and I can connect to its local AP and see the GUI. I can insert the power rocker switch and turn on the simulation and lights appear. I can toggle run and stop just fine as in assembly video 11 [1]. So far so good!

When I attempt to trigger the reset switch by shorting the middle and bottom pads of S4 (3:30 in the video), all the lights come on as expected except for LS2.

I know the LED in LS2 works from testing during assembly. I confirmed this by using a jumper to connect IC U2's pin 2 to power from capacitor C4 and the LED lights up. Resistor R42 checks out, too. Using a multimeter, it looks like the pin 2 of U2 never goes high. I swapped out the 74HC595 of U2 with the U1 and U3 and LS2 remains out. I additionally swapped around the 74HC165s and LS2 remains out. For good measure, I reflowed all IC sockets and reseated all chips, but LS2 remains out.

I found this post [2], which suggests checking continuity of pins 10-14 of all 74HC595s. Pins 10-13 all check out on all chips, but there is no continuity between any pin 14s. I am not sure if this tells me anything.

I'd appreciate any suggestions for further testing or troubleshooting!

Thanks!

[1] https://youtu.be/VrSLJicIlcU
[2] https://github.com/orgs/thehighnibble/discussions/52#discussioncomment-5806331

**Comment by nbreeden2** [comment]
Some thoughts:

The 74HC595's are 8-bit serial to parallel shift registers. The data for the LEDs is sent out as a serial stream thought the shift register from one 595 IC to the next 595 IC and so on. Let me say this a different way: Each 595 has a serial input and a serial output, the ESP32 sends the serial stream to the first 595s input in the chain, the first 595s output feeds the next 595s input and so on.

We know the serial bit stream is making it past LS2 LED as the LEDs after it work (LS3..LS7), and it sounds like the LD0 to LD7 LEDs work as well. This would indicate to me the serial data stream through all the 595s is working.

You've tested LS2 with the jumper wire and it seems to work. You've also indicated R42 looks good.

One thought: Is pin 2 on the 595 folded under the IC? Did the same thing happen on the 595 you swapped in? I've seen bad sockets cause multiple IC pins to be bent before you realize it's happening.

Some troubleshooting ideas:

Check the color bands on R42 to make sure it matches R41 and R43.

With U2 removed check the resistance of R42 vs R41 and R43. It should be very close. It not it could be a mismarked resistor.

With U2 removed take a close look at pin 2 inside the socket (not the solder joint, look at the hole in the socket). Does it look like the other pins in the socket? I've seen sockets where a pin inside the socket is damaged that isn't obvious without a close look. Pin 2 on the 595 must make good contact with the pin inside the socket. These are double swipe sockets so the pin should make contact inside a little U shaped metal receiver such that the pin is connected on both sides.

With U2 removed check the resistance from U2 Pin 2 (one lead to the pin in the socket) to R42. It should measure 0 ohms.

With U2 installed measure the resistance from U2 pin 2 (touch the shoulder on the pin up against the plastic package) to R42. It should be 0 ohms.

As a note:
The 595s drive the LEDs in a serial to parallel shift chain.
The 165s are the switch inputs, they are a parallel to serial shift chain.

Let us know what you find.
-Neil

**Reply by bohnertchris** [reply]
Thanks for the clarification of what the 595s are doing, Neil, and for the suggestions. It hadn't occurred to me that pin 2 of U2's socket could have been bad/not making contact. However, it looks as though that is not the case.

- R42's bands match up to R41 and R43.
- With U2 removed, R41, R42, and R43 all measure 2.92 K ohms
- The socket for pin 2 of U2 looks just fine and just as all other pins in the socket.
- Resistance from U2 pin2 to R42 is in fact 0 ohms with and without a 595 installed in U2.

Looking at the pins of the 595s, all are in order with nary a bend to be found.

Not sure what else it could be.

![IMG_2751](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/1bcd4217-7fe8-4ef5-9b9d-a3722796cb8d)
![IMG_2756](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/b7f9a158-073c-489c-81fc-9b8960275281)
![IMG_2757](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/2c762543-7740-41bf-a4ff-d27630cd65c5)
![IMG_1217](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/4f0e45c4-cd3d-4e28-b7c1-2b31915e529c)

**Comment by nbreeden2** [comment]
Let me give it a bit more thought. Do you have an volt/resistance meter?

**Reply by bohnertchris** [reply]
Sure do. And thanks for your continued cogitation.

**Comment by nbreeden2** [comment]
No prob, I'd like to see this working.

Can you make some voltage measurements for me? You'll need the USB cable plugged in and the board powered up. Be careful to not short anything. I've attached 3 photos with Red and Black arrows. With the meter black lead to the black arrow and the red lead to the red arrow you should see around 5 volts. I'm having you test both the Vss (GND) and Vcc traces on the board.
Test these with the IC out so directly in the socket then again with the IC in touching the pins on the IC. I know it's a bit tedious but troubleshooting often is. The first rule is always test the power supplies, what people miss is you need to test both the Vss and Vcc traces on the PCB and that power is getting to the part in question. 
Please let us know the voltages you measure. Assuming you find around 5V at the IC that voltage will be the only one I need to see.
Also please take a photo of the solder side of the PCB showing the solder joints for the IC, the cap next to it, the resistor and LED.
![IC](https://github.com/thehighnibble/thehighnibble.github.io/assets/37715892/45887442-914e-4c1d-9874-337d9fc7acb6)
![Vss](https://github.com/thehighnibble/thehighnibble.github.io/assets/37715892/1d65501e-0e88-4a24-9b1c-b8ee75f52eb4)
![Vcc](https://github.com/thehighnibble/thehighnibble.github.io/assets/37715892/fa4ee009-e72d-43b1-8706-7da044a96114)

**Reply by bohnertchris** [reply]
In summary: all voltages you asked for were around 4.52 V. I annotated the photos with the details just in case they prove useful. Please forgive my distressingly bad soldering.

I also checked U3 and U1 and was seeing essentially the same voltages. I am powering the board directly from my computer's USB, but also tried with a quality wall adapter with no observable change in behavior or voltages.
![pin8-16](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/4cb3652b-3aad-4987-84a0-2779f16e0707)
![pin8-c4](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/f02a1141-41b8-4f74-97d5-8d89618931fe)
![pin10-pin16-c4](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/961c9bc2-b954-4816-8a2d-1a77b1e04472)
![distressingly-bad-soldering](https://github.com/thehighnibble/thehighnibble.github.io/assets/38424477/4a480075-c723-499d-a9e6-f94c09a4fd12)

**Comment by nbreeden2** [comment]
Hi,
  4.5V is a definite problem. Experience has shown that low voltage causes issues with the system. A poor-quality USB cable is often the cause, I went through several before I found a good one. A powered USB hub has also helped. Many others have had this same experience. The ESP32 is a 3.3V device and has on-board regulation to generate the 3.3V from the incoming USB power. The ESP32 passes the USB power out to the rest of the system. The logic ICs really need close to 5V to work reliably. The low incoming USB voltage will also cause unexpected behavior of the ESP32, as LEDs light the current demand on the 5V will vary a lot, a poor quality USB cable cannot keep up with the current demands of the LEDs which will result in higher voltage drops across the USB cable, the 4.5V you're seeing may dip even lower which generates noise which the ESP32 is sensitive too.
  With all this said try a few USB cables and see if one resolves the issue. Try powered USB hub as well. Experience across many users has shown these are critical for the system to function well.
  
  You're soldering doesn't look bad. We all progress from it being difficult to becoming easier and more reliable over time. For the circular pads you should see full wetting of the pad (it's all covered in solder). The solder will form a little volcano shape with the component lead sticking out the top. If you end up with a solder ball either the pad isn't fully wetting (try some flux) or you have too much solder.
  With a well-shaped solder joint I gently touch the flush cutter to the component lead as if I was going to cut it high, then slide the cutter down until I feel the top of the solder, then cut. For me this tends to get a better result, practice helps with the muscle memory.
  
  Take pride in the work you've done and the skills you're learning. This is all a journey, and we all learn and improve each time we take on a project. 
  
  Let us know what you find.
-Neil

**Comment by bohnertchris** [comment]
Thanks, Neil. Sorry for going quiet, but I have been busy at work. I have managed to try a few things.

- I verified that my external power supplies (one old iPad and iPhone chargers) both are reliably putting out 5.11 V. I cut up an old USB cable and used a multimeter on the power leads. So I know at least my source is good.
- I tried the nine mini USB cables that I could find in the house connecting the iPad charger to the Z-1 replica. Measuring voltages at U2 across pins 8 and 16 showed a range from 4.30 to 4.75 V.
- I removed the ESP32-PICO-KIT and tested the voltage out pins. With my worst cable, the voltage for the 5.0 V pin would dance between 4.75 and 4.82 while the best would dance between 4.82 and 4.88. The 3.3 V pins held steady at 3.28 with both cables.

Could it be that ESP32-PICO-KIT is actually at fault? The fact that the 3.3 V line holds rock steady while the 5.0 V line fluctuates quite a bit (an never reaches 5V) makes me wonder.

I have not been able to find a powered USB hub to test yet.

Chris


PS -  I appreciate the positive feedback and encouragement. The helpful nature is why I have enjoyed engaging with the retro computing community.

**Reply by bohnertchris** [reply]
So I hit send too soon. As I was putting away my equipment after testing, I realized that I have a bunch of other micro USB power supplies  that I hadn't considered testing: Raspberry Pi power supplies.

I hooked one up to the Z-1 and wouldn't you know that across pins 8 and 16 of U2 was measuring 4.96 V? I also tested directly on the 5 V power pins of the ESP32-PICO-KIT and it held between 5.01 and 5.02 V. So now I am seeing much better voltages.

But LS2 still won't light :(

**Comment by nbreeden2** [comment]
This is excellent news! LS2 is still a mystery but we will get it figured out. I'm finally at my new place (been a LONG move) and will have my stuff delivered next week though it will take time to get my lab fully setup. At the moment I have a blank PCB and my good meter. I'll go over your photos in more detail and will compare against my board then we can go from there.  
-Neil

**Comment by nbreeden2** [comment]
One more round of measurements. Jumper the RESET switch location so the system sits in RESET so all the LEDs are in test mode (are on). You should be able to put one of the address/data toggle switches into the RESET switch location and use it to hold the system in RESET.
Now measure the voltage on each of the locations with the blue arrows. They should all be nearly the same voltage (near +5V, I'm not sure what the exact voltage will be).

If the pin that drives the LED is measures lower then the other pins swap the IC with another one and test again.

My gut is the LED is the issue but taking these steps before replacing it will help me understand the issue for future debugging efforts.

As a note:
  When you test the LEDs with the wire jumper you are applying +5V directly from the +5V power rail, a lot of current is available.
  The IC won't deliver the full +5V and can supply much less current then the +5V rail.
  If the LED requires a higher voltage/current to light vs the others this then could be root cause.

-Neil
![30](https://github.com/thehighnibble/thehighnibble.github.io/assets/37715892/41e1c0ec-d7c9-4e95-a5ab-19fb9ffb80ad)

**Reply by bohnertchris** [reply]
Thanks for the suggestion, @nbreeden2. With the board powered using the RPi power supply and S4 jumpered with a toggle switch as suggested, I measured from the negative terminal of C4 across the pins of U2 that you requested:

QA: 4.92 V
QB: 4.92
QC: 0.00
QD: 4.92
QE: 4.92
QF: 4.92
QG: 4.92
QH: 4.92

The same result was achieved by swapping the '595 in U2 with the one in U1. I also did the same measurements across all Q pins and the GND pin of U2. The only difference was that all pins but QC measured 4.91 V. QC stubbornly held at zero.

I think I see why you are suspecting the LED. Thanks for the explanation. Do you suggest swapping it out at this point or do you have some more testing suggestions?

I have been doing some traveling lately and will into the next week, so my responses will be delayed. Despite that, know that I appreciate the support. And I hope you are getting settled nicely into the new home.

Chris

**Comment by nbreeden2** [comment]
Any progress on this? Did swapping the LED work?

**Comment by bohnertchris** [comment]
Hello Neil,

I was unexpectedly called away for the past two weeks and just got back. I hope to continue with the LED swap this weekend and am eager to let you know how it goes. Fingers cross.

Thanks for the continued support and interest!

**Comment by acsidental** [comment]
Sorry my eyes aren't that great but is that a solder joint between the ground plane and pin 2 of U2? Could just be a reflection but it looks like there might be a slight join?

<img width="163" alt="image" src="https://github.com/thehighnibble/thehighnibble.github.io/assets/137387310/5cc781ed-36a9-4dfe-8810-dca5a5f1ae98">

**Reply by nbreeden2** [reply]
Good eyes! My eyes aren't what they used to be either.
I don't believe the solder flake is an issue as the LED will light when driven with the test jumper but it should be removed,
Looking closer there is a dark partial ring/shadow above the flake that might indicate a cold solder joint.
Please remove the solder flake and reflow the joint. 

<img width="163" alt="darkring" src="https://github.com/thehighnibble/thehighnibble.github.io/assets/37715892/46262d8a-0c80-4eab-ae0f-5c2f92dc4eff">

**Comment by bohnertchris** [comment]
Thanks for the critical eye on the photos, @acsidental and @nbreeden2. The solder flake was in fact just some flux residue that caught the light of my work lamp to appear silver/white. I reflowed the joint of pin two of U2 as suggested and there was no change.

I then replaced the LED in LS2 with a spare that came with the kit. After installation, I regret to say that the same problem persists. When I power up the board with the RPi power supply and hold it in reset mode with S4 jumpered, all LEDs light except LS2. Jumpering the positive pole of C4 to the positive pin of LS2 via R42 lights the LED without trouble. For sanity's sake I rechecked:

- Continuity from pin 2 of the U2 socket and R42
- Ohm rating of R42
- Swapped around '595s with those in U1 and U3

And for completeness, I also ran the same checks as you last suggested, @nbreeden2:

QA: 4.91 V
QB: 4.91
QC: 0.00
QD: 4.91
QE: 4.91
QF: 4.91
QG: 4.91
QH: 4.91

I guess we can now rule out the LED, or? Is it possible that something further up the line is causing this?

Thanks,

Chris

**Reply by acsidental** [reply]
I would suggest removing the chip and bending the QC leg out so when reinserted and retested you can test the voltage. This should tell you if the issue is before the pin or after.

**Reply by nbreeden2** [reply]
We know the IC is good as it's been swapped with other ICs. We also know the LED works when +5V is connected to the pin in the IC socket. The LED was also replaced.  The resistor was measured and was consistent with the other LEDs. 
Try this: With Pin 2 bent so it's not connected in the socket press RESET to get all LEDs lit. Using the jumper test lead connect the side of the resistor for the "bad" led (side connected to the IC socket pin 2) to another pin on the IC such as pin 3. I would expect both LEDs to be lit, they may end up dim but should lite.
Let us know what you find.

**Reply by bohnertchris** [reply]
Thanks for the continued suggestions.

I removed U2 and bent pin 2 so it is at 180° (i.e. sticking straight out from the side of the chip) so that it does not touch the socket at all when reinserted. I observed the following:

- When powering the board up with reset jumpered, LS2 is naturally out.
- When jumpering pin 3 of U2 to R42, LS2 lights up with what I perceive to be full brightness and with no effect on LS3.
- The voltage across pin 2 and pin 8 (GND) of U2 was zero.

I then bent pin 2 of the chip in U1 and then swapped the chip in U2 for that which was in U1 and vice versa. I observed the following:

- When powering the board up with reset jumpered, LO2 is naturally out.
- When jumpering pin 3 of U1 to R18, LO2 lights up with what I perceive to be full brightness and with no effect on LO3.
- The voltage across pin 2 and pin 8 (GND) of U1 was 4.89 V.

At the same time, the readings and behavior of the chip now in U2 remained unchanged:

- When powering the board up with reset jumpered, LS2 is naturally out.
- When jumpering pin 3 of U2 to R42, LS2 lights up with what I perceive to be full brightness and with no effect on LS3
- The voltage across pin 2 and pin 8 (GND) of U2 was zero.

So it seems that no matter which '595 chip that is in U2, by virtue of being in U2, it simply won't drive pin 2/LS2.

**Comment by nbreeden2** [comment]
One more attempt, can you please:
With U2 Pin 2 bent out and in RESET jumper the floating U2 Pin 2 to both sides of the resistor for LS2. When you jumper it to the side of the resistor facing the LED don't keep in lit for too long as this is bypassing the current limiting resistor.
Let us know what happens when connected to each side of the resistor.

**Reply by bohnertchris** [reply]
With pin two of U2 still bent out and the board in reset mode, all lights are on except LS2 (as expected). Jumpering pin two of U2 to either side of R42, nothing happens to either LS2 or anything else. LS2 remains dark.

**Comment by nbreeden2** [comment]
I believe we've looked at every possible hardware issue from several different directions; it seems the hardware is fine. 
This leaves us with a potential bad load of the firmware in the ESP32 which, if so, would surprise me. 
Are you located in the US / North America?
I'm a friend of Dave (The High Nibble) who can vouch for me, a potential next step would be to send the ESP32/uSD (or even the full kit) to me so I can dig deeper.
Another possibility would be for you to reimage the ESP32, while it's not difficult at this stage I believe it would need to be done using a Python script from a shell in Windows or MacOS. The kit isn't far enough along for the UI Drag/Drop or OTA methods.

**Reply by bohnertchris** [reply]
I really appreciate the offer to help further troubleshoot, @nbreeden2! I am located in Switzerland, however, and postage is nontrival. Trust me, I am sending Christmas packages to friends and family right now. As a first step, I would be more than happy to try re-imaging the ESP32 if you can point me to directions.

**Comment by nbreeden2** [comment]
Cool. I've been meaning to do a video on how to flash the ESP32 using python from the command line in Windows. I'll get this created and posted. 
There are also instructions on the HN website. 
There is a section in RED on this page......
https://thehighnibble.com/imsai8080/update/#updating-the-firmware
....that can be used to do a full command line image of the ESP32.
Imaging the uSD is straight forward, the instructions on the same page cover it.

**Comment by nbreeden2** [comment]
We finally have an explanation of the LS2 LED issue.
The kit in question is the Cromemco version which has a different firmware then the IMSAI version.
On the IMSAI version LS2 is the "STACK" LED.
On the Cromemco version LS2 is the CPU Speed LED, 2MHz or 4MHz.
A bug in the Cromemco firmware is resulting in the CPU speed indication leaking through the LED Test Code on RESET resulting in LS2 always being off. Note that the NVR (Non-Volatile RAM) as shipped sets the Z80 speed at 2MHz. This is why during the LED test step in the assembly documentation the LED isn't lit.
Dave (The High Nibble) will be addressing this in a future Cromemco firmware release.

You are safe to continue the assembly of the kit! Let us know how it goes!

**Reply by bohnertchris** [reply]
Thanks a lot for the explanation, @nbreeden2! This is great news and I am really happy to continue assembly of my Cromemco this weekend.

May I ask how you finally uncovered the culprit? Or should I just be happy that the culprit isn't actually the guy sitting behind the keyboard currently :)

**Comment by nbreeden2** [comment]
Dave and I are friends.  This issue came up and as we discussed it, he looked up your order, saw it was a Cromemco kit, he knew that LS2 was different between the two systems (STACK vs 2MHz/4MHz) and dug into the source code.
I had done a video series on assembling the IMSAI kit and Dave contacted me with an observation about one of the videos. From there a friendship evolved.
https://www.youtube.com/shadowtronblog

**Reply by bohnertchris** [reply]
Sometimes it's all about whom you know. I appreciate you two mulling it over and finally discovering the root cause. I finally completed assembly of my Z-1 kit this weekend and the results are beautiful. LS2 now lights properly indicating either 4 or 2 MHz modes.

Let me add to what I hope is an existing chorus of praise for the kit Dave has created. The build quality is amazing, it looks lovely, and is a lot of fun to play with. I would be remiss if I didn't also thank you again, @nbreeden2, and everyone else who chimed in to help me figure this out. Rediscovering old tech is fun, but it is the helpful and friendly encouragement from everyone else that really makes it a community.

Off to play with Cromix!

**Comment by nbreeden2** [comment]
You're very welcome. -Neil

**Comment by daberkow** [comment]
I had the same issue! Maybe we can add a note to the instructions and the video that this is to be expected? Maybe it was only a period of kits mailed out?

**Comment by Retro-Tech-Reboot** [comment]
I believe in the end we determined this only happens on the Cromemco emulator. 
LED LS2 for the IMSAI Emulator is STACK. On the Cromemco it's the 2MHz/4MHz indicator.  
Agreed a note in the instructions would be helpful. Hoping The High Nibble can add this.
